sdk/core/core/src/az_client_secret_credential.c

```diff
@@ -0,0 +1,237 @@
+    AZ_RETURN_IF_FAILED(az_json_get_object_member(
+        az_mut_span_to_span(response_buf), AZ_STR("access_token"), &value));
+    AZ_RETURN_IF_FAILED(az_json_value_get_string(&value, out_result));
+  }
+
+  return AZ_OK;
+}
+
+// Being given
+// "https://NNNNNNNN.vault.azure.net/secrets/Password/XXXXXXXXXXXXXXXXXXXX?api-version=7.0", gives
+// back "https://vault.azure.net" (needed for authentication).
+static AZ_NODISCARD az_result
+az_auth_get_resource_url(az_span const request_url, az_span_builder * p_builder) {
+  az_url url = { 0 };
+  if (!az_succeeded(az_url_parse(request_url, &url))) {
+    return AZ_ERROR_ARG;
+  }
+
+  az_span domains[3] = { 0 };
+  size_t const ndomains = AZ_ARRAY_SIZE(domains);
+  for (size_t i = 0; i < ndomains; ++i) {
+    if (!az_succeeded(az_host_read_domain(&url.authority.host, &domains[(ndomains - 1) - i]))) {
+      return AZ_ERROR_ARG;
+    }
+  }
+
+  // Add "https://"
+  AZ_RETURN_IF_FAILED(az_span_builder_append(p_builder, url.scheme));
+  AZ_RETURN_IF_FAILED(az_span_builder_append(p_builder, AZ_STR("://")));
+
+  for (size_t i = 0; i < (ndomains - 1); ++i) { // This loop would add "vault.azure."
+    AZ_RETURN_IF_FAILED(az_span_builder_append(p_builder, domains[i]));
+    AZ_RETURN_IF_FAILED(az_span_builder_append(p_builder, AZ_STR(".")));
+  }
+
+  // We have to do this out of the loop so that we won't append an extra "." at the end.
+  // So this expression is going to add the final "net" to an existing "https://vault.azure."
+  AZ_RETURN_IF_FAILED(az_span_builder_append(p_builder, domains[ndomains - 1]));
+
+  return AZ_OK;
+}
+
+static AZ_NODISCARD az_result az_auth_client_credentials_add_token_header(
+    az_client_secret_credential * const credential,
+    az_http_request_builder * const hrb) {
+  AZ_CONTRACT_ARG_NOT_NULL(credential);
+  AZ_CONTRACT_ARG_NOT_NULL(hrb);
+
+  az_mut_span entire_buf = (az_mut_span){
+    .begin = credential->token_credential.token,
+    .size = sizeof(credential->token_credential.token),
+  };
+  az_mut_span post_bearer = { 0 };
+  az_mut_span bearer = { 0 };
+  AZ_RETURN_IF_FAILED(az_mut_span_copy(entire_buf, AZ_STR("Bearer "), &bearer));
+  post_bearer = az_mut_span_drop(entire_buf, bearer.size);
+
+  az_span_builder auth_url_builder = az_span_builder_create(post_bearer);
+  AZ_RETURN_IF_FAILED(az_auth_get_resource_url(az_mut_span_to_span(hrb->url), &auth_url_builder));
+  az_span const auth_url = az_span_builder_result(&auth_url_builder);
+
+  az_span token = { 0 };
+  AZ_RETURN_IF_FAILED(az_auth_get_token(
+      credential, auth_url, az_mut_span_drop(post_bearer, auth_url.size), &token));
+
+  az_mut_span unused;
+  AZ_RETURN_IF_FAILED(az_mut_span_move(post_bearer, token, &unused));
+
+  AZ_RETURN_IF_FAILED(az_http_request_builder_append_header(
+      hrb,
+      AZ_STR("authorization"),
+      (az_span){ .begin = bearer.begin, .size = bearer.size + token.size }));
+
+  return AZ_OK;
+}
+
+AZ_NODISCARD az_result az_client_secret_credential_init(
```

Comment:
probably this function should be the first one in this file..  order is not important, but usually people expect to see inits as the first function IMO

Reply:
Could also be INLINE ??

Reply:
It needs to be at the end, because it depends on the previous declarations... And we can't make it inline, because it initializes .func with a pointer to a static function, which is not visible outside of a compilation unit (c file), and it is better not to declare global functions without necessity.

I made the other two inline.